cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -1422,6 +1424,33 @@ toTentativeEpochInfo (EraHistory _ interpreter) =
     $ hoistEpochInfo (first (Text.pack . show) . runExcept)
     $ Consensus.interpreterToEpochInfo (Consensus.unsafeExtendSafeZone interpreter)
 
+
+-- | Get slot number for timestamp, or an error if the UTC timestamp is before 'SystemStart'
+utcTimeToSlotNo
+  :: SocketPath
+  -> AnyConsensusModeParams
+  -> NetworkId
+  -> UTCTime
+  -> ExceptT ShelleyQueryCmdError IO SlotNo
+utcTimeToSlotNo (SocketPath sockPath) (AnyConsensusModeParams cModeParams) network utcTime = do
+  let localNodeConnInfo = LocalNodeConnectInfo cModeParams network sockPath
+  case consensusModeOnly cModeParams of
+    CardanoMode -> do
+      (systemStart, eraHistory) <- executeLocalStateQueryExpr' localNodeConnInfo $
+        (,) <$> queryExpr' QuerySystemStart
+            <*> queryExpr' (QueryEraHistory CardanoModeIsMultiEra)
+      let relTime = toRelativeTime systemStart utcTime
+      hoistEither $ Api.getSlotForRelativeTime relTime eraHistory & first ShelleyQueryCmdPastHorizon
+    mode -> left . ShelleyQueryCmdUnsupportedMode $ AnyConsensusMode mode
+  where
+    executeLocalStateQueryExpr' localNodeConnInfo =
+      ExceptT
+      . fmap (join . first ShelleyQueryCmdAcquireFailure)
+      . executeLocalStateQueryExpr localNodeConnInfo Nothing
+      . runExceptT
+    queryExpr' = withExceptT ShelleyQueryCmdUnsupportedNtcVersion . ExceptT . queryExpr
```

Comment:
Adding type signatures to `where` functions would add visibility to what the types of these functions are.